File: solutions/5. Longest Palindromic Substring.js
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function(s) {
    if(!s || s.length === 0) return s;
    let n = s.length;
    let pal = [...Array(n+1)].map(x=>Array(n+1).fill(false));
    let ans = ''
    for (let i = 0; i < n; ++i) { 
        pal[i][i] = true;
        ans = s.substr(i,1);
    }
    let max = 1;
    for (let i = 0; i < n-1; ++i)  {
        if (s.charAt(i) === s.charAt(i + 1)) {
            pal[i][i+1] = true;
            ans = s.substr(i, 2);
            max = 2;
        }
        
    }
    for (let k = 3; k <= n; ++k) {
        for (let i = 0; i < n - k + 1; ++i) {
            let j = i + k - 1;
            if(pal[i+1][j-1] && s.charAt(i) === s.charAt(j)) {
                pal[i][j] = true;
                if (max < k) {
                    max = k;
                    ans = s.substr(i, k)
                }
            }
        }
    }
    return ans;
    
};

File: solutions/5. Longest Palindromic Substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        string str;
        str += "@";
        for(int i = 0; i < s.length(); ++i) {
            str.push_back('#');
            str.push_back(s[i]);
        }
        str +="#$";
        int n = str.length();
        int dp[n] = {0};
        int c = 1, i = 1, r = 1;
        int start = 1, max = 1;
        while(i+1 < n) {
            int mirr = c - (i-c);
            if(r > i) {
                dp[i] = min(dp[mirr], r-i); 
            }
            while(str[i+(dp[i]+1)] == str[i-(1+dp[i])])
                dp[i]++;
            if(dp[i] > max) {
                start = i;
                max = dp[i];
            }
            if(r < i + dp[i]) {
                c = i;
                r = i + dp[i];
            }
            i++;
        }
        return s.substr((start-1-max)/2, max);
    }
};
